### PR TITLE
Puma ruby webserver integration

### PR DESCRIPTION
**Pull request by plasticine**
### What does this PR do?

This patch implements a [Puma](https://puma.io/) webserver integration, leveraging the in-built stats endpoint that Puma supports.

### Motivation

It’s extremely valuable when running Puma in production to be able to monitor the webserver backlog and capacity values.

### Review checklist

- [x] PR has a [meaningful title](https://github.com/DataDog/integrations-core/blob/master/CONTRIBUTING.md#pull-request-title) or PR has the `no-changelog` label attached
- [x] Feature or bugfix has tests
- [x] Git history is clean
- [ ] If PR impacts documentation, docs team has been notified or an issue has been opened on the [documentation repo](https://github.com/DataDog/documentation/issues/new)

### Additional Notes

Anything else we should know when reviewing?

### CODE REVIEW

**Comment by hithwen** [comment]
Hallo @plasticine,
This PR is a draft so we are not reviewing it. I just want to double check that this is intended.

**Comment by plasticine** [comment]
@hithwen Ah—my mistake, I didn’t realise! 😅 I’d love a review if it’s of interest. Will mark it as ready for review.

File: puma/README.md
Comment (ChristineTChen) [inline puma/README.md:1]:
```suggestion
# Agent Check: Puma
```

File: puma/README.md
Comment (ChristineTChen) [inline puma/README.md:5]:
```suggestion
This check monitors [Puma][1] through the Datadog Agent.
```

File: puma/README.md
Comment (ChristineTChen) [inline puma/README.md:13]:
```suggestion
To install the Puma check on your host:
```

File: puma/README.md
Comment (ChristineTChen) [inline puma/README.md:48]:
```suggestion
[1]: https://puma.io/
```
please confirm if that is the correct site

File: puma/README.md
Comment (ChristineTChen) [inline puma/README.md:5]: Can we include a brief description of Puma and what this integration can monitor?

File: puma/README.md
Comment (ChristineTChen) [inline puma/README.md:38]:
```suggestion
**puma.connection**: Returns `CRITICAL` if the Agent is unable to connect to the monitored Puma instance. Returns `OK` otherwise.
```

File: puma/datadog_checks/puma/puma.py
Comment (ChristineTChen) [inline puma/datadog_checks/puma/puma.py:24]:
```suggestion
        control_url = instance.get('control_url')
        if control_url is None:
            raise ConfigurationError('Puma instance missing "control_url" value')

        tags = instance.get('tags', [])
        response, content_type, version = self._perform_service_check(instance, control_url)
        metrics = self._extract_metrics(response)
```

Let's set `control_url` to a variable as it's used more than once

File: puma/datadog_checks/puma/puma.py
Comment (ChristineTChen) [inline puma/datadog_checks/puma/puma.py:81]: 💯 thanks for using!

File: puma/assets/configuration/spec.yaml
Comment (ChristineTChen) [inline puma/assets/configuration/spec.yaml:15]:
```suggestion
    - template: instances/http
    - template: instances/default

```

let's include the templates for http options

File: puma/assets/configuration/spec.yaml
Comment (ChristineTChen) [inline puma/assets/configuration/spec.yaml:6]:
```suggestion
    options:
        - template: init_config/http
        - template: init_config/default

```

let's include the templates for http options and default options

File: puma/manifest.json
Comment (ChristineTChen) [inline puma/manifest.json:7]:
```suggestion
  "metric_to_check": "puma.workers",
```

File: puma/manifest.json
Comment (ChristineTChen) [inline puma/manifest.json:9]: Let's incude a short description of what this integration monitors

File: puma/manifest.json
Comment (ChristineTChen) [inline puma/manifest.json:19]:
```suggestion
    "web"
```
let's include some categories. They can be found here https://docs.datadoghq.com/integrations/

File: puma/metadata.csv
Comment (ChristineTChen) [inline puma/metadata.csv:1]: Can we include the metric name, type, unit_name (if applicable), description, integration and short name of the metrics submitted?

```suggestion
metric_name,metric_type,interval,unit_name,per_unit_name,description,orientation,integration,short_name
puma.workers,gauge,,unit,,Number of puma workers,0,puma,puma worker
```

^example of the `puma.workers` metric

File: puma/manifest.json
Comment (ChristineTChen) [inline puma/manifest.json:2]:
```suggestion
  "display_name": "Puma",
```

**Review by ChristineTChen — CHANGES_REQUESTED** [review]
Hi @plasticine! Thanks for opening this PR for the puma integration.

Can you add the integration and yourself to the CODEOWNERS file here: https://github.com/DataDog/integrations-extras/blob/master/.github/CODEOWNERS

The CI is currently failing due to the integration missing for the CI configuration. You can fix it by running `ddev -x validate ci --fix` on your branch and pushing the changes. It will add puma to the .azure_pipelines files.

I've also left some comments on to clarify wording or missing from the integration scaffolding.

**Comment by plasticine** [comment]
@ChristineTChen Amazing — thank you very much! I’ll get these cleaned up! 👍

File: puma/assets/configuration/spec.yaml
Comment (ChristineTChen) [inline puma/assets/configuration/spec.yaml:1]:
```suggestion
name: Puma
```

Comment (ChristineTChen) [reply puma/assets/configuration/spec.yaml:1]: Please also run `ddev -x validate config --sync` for the conf.yaml.example to get updated and reflect the changes added in this spec.yaml

**Review by ChristineTChen — CHANGES_REQUESTED** [review]
Looks great! I've also requested the docs team to review the documentation related files.

You may need to pull master and rebase your branch. There may be some CI failures unrelated to this integration that will block this PR.

File: puma/assets/configuration/spec.yaml
Comment (plasticine) [reply puma/assets/configuration/spec.yaml]: Done, and rebased/pushed 👍

**Comment by hithwen** [comment]
Hallo, can you please run `ddev validate config -s` and update the changes. You may need to update your current `ddev` version to the latests available.

**Comment by plasticine** [comment]
> Hallo, can you please run `ddev validate config -s` and update the changes. You may need to update your current `ddev` version to the latests available. 

Ahh, I have run it a few times, but perhaps I have an old version. Will upgrade and try again!

File: puma/assets/configuration/spec.yaml
Comment (kayayarai) [inline puma/assets/configuration/spec.yaml:14]:
```suggestion
      description: Control URL for your Puma instance
```

File: puma/manifest.json
Comment (kayayarai) [inline puma/manifest.json:9]:
```suggestion
  "short_description": "A fast, concurrent web server for Ruby and Rack",
```

File: puma/manifest.json
Comment (kayayarai) [inline puma/manifest.json:17]:
```suggestion
  "public_title": "Datadog-Puma Integration",
```

File: puma/README.md
Comment (kayayarai) [inline puma/README.md:9]:
```suggestion
Follow the instructions below to install and configure this check for an Agent running on a host. For containerized environments, see the [Autodiscovery Integration Templates][3] for guidance on applying these instructions.
```

File: puma/README.md
Comment (kayayarai) [inline puma/README.md:15]:
```suggestion
1. Install the [developer toolkit][4] on any machine.
```

File: puma/README.md
Comment (kayayarai) [inline puma/README.md:50]:
```suggestion
[3]: https://docs.datadoghq.com/agent/kubernetes/integrations/
[4]: https://docs.datadoghq.com/developers/integrations/new_check_howto/#developer-toolkit
[5]: https://app.datadoghq.com/account/settings#agent
```

File: puma/README.md
Comment (kayayarai) [inline puma/README.md:17]:
```suggestion
3. [Download the Datadog Agent][5].
```

File: puma/README.md
Comment (kayayarai) [inline puma/README.md:51]:
```suggestion
[6]: https://github.com/DataDog/integrations-extras/blob/master/puma/datadog_checks/puma/data/conf.yaml.example
```

File: puma/README.md
Comment (kayayarai) [inline puma/README.md:22]:
```suggestion
1. Edit the `puma.d/conf.yaml` file, in the `conf.d/` folder at the root of your Agent's configuration directory to start collecting your Puma performance data. See the [sample puma.d/conf.yaml][6] for all available configuration options.
```

File: puma/README.md
Comment (kayayarai) [inline puma/README.md:52]:
```suggestion
[7]: https://docs.datadoghq.com/agent/guide/agent-commands/#start-stop-and-restart-the-agent
```

File: puma/README.md
Comment (kayayarai) [inline puma/README.md:24]:
```suggestion
2. [Restart the Agent][7].
```

File: puma/README.md
Comment (kayayarai) [inline puma/README.md:53]:
```suggestion
[8]: https://docs.datadoghq.com/agent/guide/agent-commands/#agent-status-and-information
```

File: puma/README.md
Comment (kayayarai) [inline puma/README.md:28]:
```suggestion
[Run the Agent's status subcommand][8] and look for `puma` under the Checks section.
```

File: puma/README.md
Comment (kayayarai) [inline puma/README.md:54]:
```suggestion
[9]: https://github.com/DataDog/integrations-extras/blob/master/puma/metadata.csv
```

File: puma/README.md
Comment (kayayarai) [inline puma/README.md:34]:
```suggestion
See [metadata.csv][9] for a list of metrics provided by this check.
```

File: puma/README.md
Comment (kayayarai) [inline puma/README.md:42]:
```suggestion
Puma does not include any events.
```

File: puma/README.md
Comment (kayayarai) [inline puma/README.md:55]:
```suggestion
[10]: https://docs.datadoghq.com/help/
```

File: puma/README.md
Comment (kayayarai) [inline puma/README.md:46]:
```suggestion
Need help? Contact [Datadog support][10].
```

**Review by kayayarai — CHANGES_REQUESTED** [review]
Some docs review feedback for you. Thanks!

**Comment by plasticine** [comment]
Thanks @kayayarai & @hithwen! I’ve pulled in all the docs changes — awesome! Have also re-validated. You were right @hithwen I was back on `ddev` `3.x`

File: puma/manifest.json
Comment (ChristineTChen) [inline puma/manifest.json:31]:
```suggestion
    "service_checks": "assets/service_checks.json",
    "metrics_metadata": "metadata.csv"
```
Recently added this field to all integration manifest.json

**Review by ChristineTChen — APPROVED** [review]
Looks great! 💯

**Review by kayayarai — APPROVED** [review]

**Pull request merged**
